README.md

```diff
@@ -1,16 +1,695 @@
+
+# 2 Roles
+
+Invocation adds two new roles to UCAN: invoker and executor. The existing UCAN delegator and delegate principals MUST persist to the invocation.
+
+| UCAN Field | Delegation                             | Invocation                      |
+|------------|----------------------------------------|---------------------------------|
+| `iss`      | Delegator: transfer authority (active) | Invoker: request task (active)  |
+| `aud`      | Delegate: gain authority (passive)     | Executor: perform task (active) |
+
+## 2.1 Invoker
+
+The invoker signals to the executor that a task associated with a UCAN SHOULD be performed.
+
+The invoker MUST be the UCAN delegator. Their DID MUST be authenticated in the `iss` field of the contained UCAN.
+
+## 2.2 Executor
+
+The executor is directed to perform some task described in the UCAN by the invoker.
+
+The executor MUST be the UCAN delegate. Their DID MUST be set the in `aud` field of the contained UCAN.
+
+# 3 Envelope
+
+The invocation envelope is a thin wrapper around a UCAN that conveys that all of the contained tasks SHOULD be performed.
+
+All of the roles from the referenced UCANs MUST persist to the invocation per the [roles table](#2-roles).
+
+The invocation wrapper MUST be signed by the same principal that issued the UCAN.
+
+## 3.1 Fields
+
+| Field | Type                     | Description                                            | Required | Default |
+|-------|--------------------------|--------------------------------------------------------|----------|---------|
+| `v`   | `SemVer`                 | SemVer of the UCAN invocation this schema (v0.1.0)     | Yes      |         |
+| `prf` | `[&UCAN]`                | UCANs that supply the authority to run this invocation | Yes      |         |
+| `run` | `"*" or {String : Task}` | Which UCAN capabilities to run                         | Yes      |         |
+| `nnc` | `String`                 | A unique nonce, to distinguish each invocation         | Yes      |         |
+| `ext` | `Any`                    | Non-normative extended fields                          | No       | `null`  |
+    
+### 3.1.1 Version
+
+The `v` field MUST contain the version of the invocation object  schema.
+
+### 3.1.2 Proofs
+
+The `prf` field MUST contain CIDs pointing to the UCANs that provide the authority to run these tasks. The elements of this array MUST be sorted in ascending numeric order. Restricting the outmost UCANs to only the authority required for the current invocation is RECOMMENDED.
+
+### 3.1.3 Run Capabilities
+
+The `run` field MUST reference the tasks contained in the UCAN are to be run during the invocation. To run all tasks in the underlying UCAN, the `"*"` value MUST be used. If only specific tasks (or [pipelines](#5-promise-pipelining)) are intended to be run, then they MUST be prefixed with an arbitrary label and treated as a UCAN attenuation: all tasks MUST be backed by a matching capability of equal or greater authority.
```

Comment:
Does tasks correspond to capabilities or UCANs ? I'm also starting to think that `*` may be error prone and perhaps it's not worth having it.

For what it's worth my former desire to use `*` was misguided, our current invocation is actually a tuple of CIDs (to UCANs that have single capability delegation each).

Reply:
> Does tasks correspond to capabilities or UCANs ? I'm also starting to think that * may be error prone and perhaps it's not worth having it.

Yes, exactly

> For what it's worth my former desire to use * was misguided, our current invocation is actually a tuple of CIDs (to UCANs that have single capability delegation each).

Ah, that's really good to know!

A few days I added and then quickly reverted this:

```js
{
  // ...
  "run": [
    "bafyUcan1",
    "bafyUcan2"
  ]
}
```

Which has some other nice knock-on effects 🤔 I wonder if we could extend this to what I think you're saying:

```js
{
  // ...
  "run": [
    ["bafyUcan1", "bafyHashOfUriAndAbility1"],
    ["bafyUcan2", "bafyHashOfUriAndAbility2"]
  ]
}
```

or for consistency

```js
{
  // ...
  "run": {
    "foo": ["bafyUcan1", "bafyHashOfUriAndAbility1"],
    "bar": ["bafyUcan2", "bafyHashOfUriAndAbility2"]
  }
}
```

Both of which look a whole lot like the promise pointer:

```json
{"ucan/promise": ["bafyUcan1", "bafyHashOfUriAndAbility1"]}
```

Maybe we factor out the pointer, and let it be used generically in the `run` and promises?

Reply:
I like the idea of having canonical form for referring to result of the task. That is what I was talking about when I was suggesting maybe whole promise sugar can be rendered obsolete, if we have that than we can refer to it either as a result I want get back or as an input of the task I want result from

Reply:
As an additional context tuple if CIDs was a compromise to ship something soon and then iterate. Originally I was going after graphql like query syntax so you could name results and provide selectors to avoid passing back unnecessary data, which is more inline what’s in the spec.

I might try and resurrect branch that was headed that direction to evaluate this spec